Example target files

MarcKarasek opened the pull request:
This dir is to be used as a template for development
of a target for the riscv architectural tests.

neelgala commented on riscv-target/example-target/compliance_model.h:
```diff
@@ -0,0 +1,166 @@
+// Code below will copy from the rom area to ram the 
+// data.strings and .data sections to ram.
+// Use linksplit.ld 
+#define RVTEST_TARGET_INIT \
```
I think should be renamed to RVMODEL_BOOT which is already used in tests before the RVTEST_CODE_BEGIN. Used for target specific initializatio sequences. Also in general all target specific macros should be prefixed RVMODEL_ and not RVTEST_

MarcKarasek replied:
Since RVMODEL_BOOT comes before the RVTEST_CODE_BEGIN this would not work.  If we moved the RVNODEL_BOOT to after the RVTEST_CODE_BEGIN then it would.  

Currently there is nothing defined for RVMODEL_BOOT.  So I see no reason not to move the macro in the test code and rename what I have added to RVMODEEL_BOOT.

neelgala replied:
Maybe I am missing it completely.. but why does it have to be after RVTEST_CODE_BEGIN ? 
Shouldn't all initializtion of the target and ios happen before entering the test. 

One issue with having it after, is we start capturing coverage of the test between RVTEST_CODE_BEGIN and RVTEST_CODE_END. So everything in RVMODEL_TARGET_INIT is going to contribute to coverage. Also the RVTEST_CODE_BEGIN initializes the registers to specific values which will be lost if RVMODEL_TARGET_INIT is after it.

MarcKarasek replied:
It has to be after the RVTEST_CODE_BEGIN because this is where the code starts executing.  
It is where :   
.globl rvtest_start;                                                  \
 rvtest_start:

is defined.  So this is the entry point for code execution.  So any init code needed by a target must be after this in order for it to execute.

neelgala replied:
A target always uses a custom linker script. So you can also define the entry point in the linker script. Basically in RVMODEL_BOOT you could have a label : rvmodel_boot_start and use this as the entry point in the linker script. Have your boot-up code and then simply fall into RVTEST_CODE_BEGIN. 

Essentially replace rvtest_start in this [line](https://github.com/incoresemi/riscv-compliance/blob/34d979eaf859d8608adaa4781235f327bc51a5f8/riscv-target/spike/link.ld#L2) with rvmodel_boot_start.

So then you don't need RVMODEL_TARGET_INIT.

neelgala commented on riscv-target/example-target/README.txt:
```diff
@@ -0,0 +1,49 @@
+// Code below will copy from the rom area to ram the 
+// data.strings and .data sections to ram.
+// Use linksplit.ld 
+#define RVTEST_TARGET_INIT \
```
rename to RVMODEL_BOOT

neelgala commented on riscv-target/example-target/README.txt:
```diff
@@ -0,0 +1,49 @@
+// Any specific target init code should be put here
+// Code for one monolithic ram area
+// Use linkmono.ld 
+#define RVTEST_TARGET_INIT \
```
rename to RVMODEL_BOOT

neelgala commented on riscv-target/example-target/compliance_model.h:
```diff
@@ -0,0 +1,166 @@
+// _I = Immediate
+// This code will check a test to see if the results 
+// match the expected value.
+#define RVMODEL_CHECK_TEST(_SP, _R, _I)                                 \
```
Where should this macro be included in the compliance_test.h ? 
If you are introducing a new macro can you also add the description in the test-format-spec [here](https://github.com/incoresemi/riscv-compliance/blob/dev/spec/TestFormatSpec.adoc#43-the-test-structure-of-an-architectural-test)

allenjbaum replied:
I don't see how this will be usable unless you customize the test in the riscv-test-suite directory; where is it called?
I think this can simply be a version of the assert macro; that is already custom to the model, so you can use it as an actual assertion, or ignore the assertion value and just print the operands and result., have it halt or continue, etc, etc. or pass it a parameter to do it either way. This adds no functionality that I can see, and to be useful it would have to be precompiled into every test,

Unless I am mistaken: a test cannot execute any macro that is not already explicit in the test itself or is called by an existing model specific macro in the test.

MarcKarasek replied:
It would only have to be added in one place at the end of the TEST_CASE macro(s) that need this in compliance_test.h

I will add to this PR the change for the TEST_CASE macro.

neelgala commented on riscv-target/example-target/device/rv32i_m/Makefile.include:
```diff
@@ -0,0 +1,31 @@
+TARGET_SIM   ?= <sim runtime>
```
I would suggest using the template for spike available [here](https://github.com/incoresemi/riscv-compliance/blob/dev/riscv-target/spike/device/rv32i_m/I/Makefile.include) . The link points to commands that were suggest by Bill (seagate) to improve failure responses. I would like to keep those and the command to intermix the disassembly with source code as well.

MarcKarasek replied:
that template is out of date,,

neelgala commented on riscv-target/example-target/device/rv64i_m/Makefile.include:
```diff
@@ -0,0 +1,31 @@
+TARGET_SIM   ?= <sim runtime>
```
use template from [here](https://github.com/incoresemi/riscv-compliance/blob/dev/riscv-target/spike/device/rv32i_m/I/Makefile.include)

MarcKarasek replied:
same as for 32 bit targets.

MarcKarasek commented on riscv-target/example-target/device/rv64i_m/Makefile.include:
```diff
@@ -4,28 +4,44 @@ ifeq ($(shell command -v $(TARGET_SIM) 2> /dev/null),)
+	$$(RISCV_OBJCOPY) -O binary $$@ $$@.bin
+
+OBJ_CMD = $$(RISCV_OBJDUMP) $$@ -D > $$@.objdump; \
+		  $$(RISCV_OBJDUMP) $$@ --source-comment="// " > $$@.debug \
```
--source-comment is not supported in the gcc I am using .. 
riscv64-unknown-elf-gcc (GCC) 9.2.0

neelgala replied:
yea.. it works on Binutils 2.34+ . But it works if you use just ``--source`` even on older versions of binutils. I get the following in the objdump:
```
0000000080000398 <inst_0>:                                                                          
//                                                                                                  
// inst_0:                                                                                          
// // rs2 == rd != rs1, rs1==x22, rs2==x3, rd==x3, rs1_val == (-2**(xlen-1)), rs1_val < 0 and rs2_val < 0, rs1_val == -9223372036854775808,    rs1_val != rs2_val, rs2_val == -1099511627777 
// // opcode: add ; op1:x22; op2:x3; dest:x3; op1val:-9223372036854775808;  op2val:-1099511627777   
// TEST_RR_OP(add, x3, x22, x3, 0x7ffffeffffffffff, -9223372036854775808, -1099511627777, x7, 0, x13) 
    80000398: fff00b1b            addiw s6,zero,-1                                                  
    8000039c: 03fb1b13            slli  s6,s6,0x3f                                                  
    800003a0: fff0019b            addiw gp,zero,-1                                                  
    800003a4: 02819193            slli  gp,gp,0x28                                                  
    800003a8: fff18193            addi  gp,gp,-1                                                    
    800003ac: 003b01b3            add gp,s6,gp                                                      
    800003b0: 0033b023            sd  gp,0(t2)              
```

I'll push in these changes

MarcKarasek replied:
I have binutils 2.32 and neither --source or --source-comment works.  The --source does not fail the compile but I see no source code in the output.  The --source-comment fails compile.

Unless we are going to force people onto a certain version or later of the tools, I would recommend using just the --source as this causes no damage with older tools.

MarcKarasek replied:
Just an FYI:  Mandating a certain version or later of tools needs to be done soon if needed.  We may already be too late for this.  In my experience moving from one toolchain to a newer version is a hard sell at times.  Unless there is a mandate to do it,  a vendor supplied toolchain is one example.  But if you build your own, then upgrading to a newer version can be something a company does not want to do.  Right or wrong they see it as a untested variable in there build process.

MarcKarasek replied:
Please do not push these yet until we make a decision on how to handle mandating a certain version or later of the tools.

neelgala replied:
you have to compile the code with the ``-g``. This adds debug symbols which can be used by ``--source`` in the objdump command to mix source code in objdump. Are you using the ``-g`` in your gcc command ?

``--source`` is a pretty old command. ``--source-comment`` is recently added and I will probably resort to the former one to avoid the whole toolchain issue.

MarcKarasek commented:
This means that someone has to know if they add code for init they also
need to change the linker script.  I would rather move the RVMODEL_BOOT to
after the RVTEST_CODE_BEGIN and use it for any target specific init that
needs to be done.   I do not quite understand whay RVMODEL_BOOT is supposed
to be used for but no one is using it..

From the spec

  `RVMODEL_BOOT`::
    - contains boot code for the test-target; may include emulation code or
trap stub. If the
      test-target enforces alignment or value restrictions on the mtvec
csr, it is required that
      this macro sets the value of mtvec to a region which is readable and
writable by the machine
      mode.

The only way this boot code will be executed is if it is below the
RVTEST_CODE_BEGIN macro.




On Tue, Nov 10, 2020 at 11:20 AM Neel Gala <notifications@github.com> wrote:

> *@neelgala* commented on this pull request.
> ------------------------------
>
> In riscv-target/example-target/compliance_model.h
> <https://github.com/incoresemi/riscv-compliance/pull/1#discussion_r520691808>
> :
>
> > +//RV_COMPLIANCE_DATA_END
> +#define RVMODEL_DATA_END                                                      \
> +  .global end_signature; end_signature:
> +
> +//RVTEST_IO_INIT
> +// Add code here to initialize any IO for output support
> +#define RVTEST_IO_INIT
> +
> +//RVTEST_TARGET_INIT
> +// Any specific target init code should be put here
> +
> +// For code that has a split rom/ram area
> +// Code below will copy from the rom area to ram the
> +// data.strings and .data sections to ram.
> +// Use linksplit.ld
> +#define RVTEST_TARGET_INIT \
>
> A target always uses a custom linker script. So you can also define the
> entry point in the linker script. Basically in RVMODEL_BOOT you could have
> a label : rvmodel_boot_start and use this as the entry point in the linker
> script. Have your boot-up code and then simply fall into RVTEST_CODE_BEGIN.
>
> Essentially replace rvtest_start in this line
> <https://github.com/incoresemi/riscv-compliance/blob/34d979eaf859d8608adaa4781235f327bc51a5f8/riscv-target/spike/link.ld#L2>
> with rvmodel_boot_start.
>
> So then you don't need RVMODEL_TARGET_INIT.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/incoresemi/riscv-compliance/pull/1#discussion_r520691808>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD72ZM533VADHHOPOLYI6LLSPFR5LANCNFSM4TM53OMA>
> .
>


-- 
/****************
Marc Karasek
ASM T648
Herder of Cats
****************/

allenjbaum commented:
I was under the impression that rvmodel_boot is new, and that it's been
done a variety of different ways in the past; this is just putting it
somewhere specific, so it's no surprise that no one is suing it now.


On Tue, Nov 10, 2020 at 2:07 PM Marc Karasek <notifications@github.com>
wrote:

> This means that someone has to know if they add code for init they also
> need to change the linker script. I would rather move the RVMODEL_BOOT to
> after the RVTEST_CODE_BEGIN and use it for any target specific init that
> needs to be done. I do not quite understand whay RVMODEL_BOOT is supposed
> to be used for but no one is using it..
>
> From the spec
>
> `RVMODEL_BOOT`::
> - contains boot code for the test-target; may include emulation code or
> trap stub. If the
> test-target enforces alignment or value restrictions on the mtvec
> csr, it is required that
> this macro sets the value of mtvec to a region which is readable and
> writable by the machine
> mode.
>
> The only way this boot code will be executed is if it is below the
> RVTEST_CODE_BEGIN macro.
>
>
>
>
> On Tue, Nov 10, 2020 at 11:20 AM Neel Gala <notifications@github.com>
> wrote:
>
> > *@neelgala* commented on this pull request.
> > ------------------------------
> >
> > In riscv-target/example-target/compliance_model.h
> > <
> https://github.com/incoresemi/riscv-compliance/pull/1#discussion_r520691808
> >
> > :
> >
> > > +//RV_COMPLIANCE_DATA_END
> > +#define RVMODEL_DATA_END \
> > + .global end_signature; end_signature:
> > +
> > +//RVTEST_IO_INIT
> > +// Add code here to initialize any IO for output support
> > +#define RVTEST_IO_INIT
> > +
> > +//RVTEST_TARGET_INIT
> > +// Any specific target init code should be put here
> > +
> > +// For code that has a split rom/ram area
> > +// Code below will copy from the rom area to ram the
> > +// data.strings and .data sections to ram.
> > +// Use linksplit.ld
> > +#define RVTEST_TARGET_INIT \
> >
> > A target always uses a custom linker script. So you can also define the
> > entry point in the linker script. Basically in RVMODEL_BOOT you could
> have
> > a label : rvmodel_boot_start and use this as the entry point in the
> linker
> > script. Have your boot-up code and then simply fall into
> RVTEST_CODE_BEGIN.
> >
> > Essentially replace rvtest_start in this line
> > <
> https://github.com/incoresemi/riscv-compliance/blob/34d979eaf859d8608adaa4781235f327bc51a5f8/riscv-target/spike/link.ld#L2
> >
> > with rvmodel_boot_start.
> >
> > So then you don't need RVMODEL_TARGET_INIT.
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/incoresemi/riscv-compliance/pull/1#discussion_r520691808
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AD72ZM533VADHHOPOLYI6LLSPFR5LANCNFSM4TM53OMA
> >
> > .
> >
>
>
> --
> /****************
> Marc Karasek
> ASM T648
> Herder of Cats
> ****************/
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/incoresemi/riscv-compliance/pull/1#issuecomment-724995909>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHPXVJUXYLSL6ASDG6ZJF53SPG2QVANCNFSM4TM53OMA>
> .
>

MarcKarasek commented:
It is in the wrong place anyway. It should be after the RVTEST_CODE_BEGIN.
If this is supposed to be code run before the tests (init or setup vectors,
etc..) then it should be after the code entry point. Which is in
RVTEST_CODE_BEGIN.

To Neel point about having them designate a new entry point in the code in
the RVMODEL_BOOT macro and in the linker script (which is target
dependent).  The less complicated we can make how to add a target the
better.  IMO linker scripts are the least used and most misunderstood file
period.  Most people do not even bother with them and just go with the
defaults.  It is really only when you are on bare metal that they become
useful and needed to package things up.

Case in point I do not think anyone on the call the other day had a clue
how to use a linker script to package a binary image that has a data
section that at runtime is somewhere else.  Hence my email and
explanation.
BTW: Should this example linker file and text go into the spec/docs?

I have moved the init code I have into RVMODEL_BOOT and moved the
RVMODEL_BOOT after RVTEST_CODE_BEGIN so now it looks like this in all the
source.

RVTEST_CODE_BEGIN
RVMODEL_BOOT

Marc

On Tue, Nov 10, 2020 at 5:27 PM Allen Baum <notifications@github.com> wrote:

> I was under the impression that rvmodel_boot is new, and that it's been
> done a variety of different ways in the past; this is just putting it
> somewhere specific, so it's no surprise that no one is suing it now.
>
>
> On Tue, Nov 10, 2020 at 2:07 PM Marc Karasek <notifications@github.com>
> wrote:
>
> > This means that someone has to know if they add code for init they also
> > need to change the linker script. I would rather move the RVMODEL_BOOT to
> > after the RVTEST_CODE_BEGIN and use it for any target specific init that
> > needs to be done. I do not quite understand whay RVMODEL_BOOT is supposed
> > to be used for but no one is using it..
> >
> > From the spec
> >
> > `RVMODEL_BOOT`::
> > - contains boot code for the test-target; may include emulation code or
> > trap stub. If the
> > test-target enforces alignment or value restrictions on the mtvec
> > csr, it is required that
> > this macro sets the value of mtvec to a region which is readable and
> > writable by the machine
> > mode.
> >
> > The only way this boot code will be executed is if it is below the
> > RVTEST_CODE_BEGIN macro.
> >
> >
> >
> >
> > On Tue, Nov 10, 2020 at 11:20 AM Neel Gala <notifications@github.com>
> > wrote:
> >
> > > *@neelgala* commented on this pull request.
> > > ------------------------------
> > >
> > > In riscv-target/example-target/compliance_model.h
> > > <
> >
> https://github.com/incoresemi/riscv-compliance/pull/1#discussion_r520691808
> > >
> > > :
> > >
> > > > +//RV_COMPLIANCE_DATA_END
> > > +#define RVMODEL_DATA_END \
> > > + .global end_signature; end_signature:
> > > +
> > > +//RVTEST_IO_INIT
> > > +// Add code here to initialize any IO for output support
> > > +#define RVTEST_IO_INIT
> > > +
> > > +//RVTEST_TARGET_INIT
> > > +// Any specific target init code should be put here
> > > +
> > > +// For code that has a split rom/ram area
> > > +// Code below will copy from the rom area to ram the
> > > +// data.strings and .data sections to ram.
> > > +// Use linksplit.ld
> > > +#define RVTEST_TARGET_INIT \
> > >
> > > A target always uses a custom linker script. So you can also define the
> > > entry point in the linker script. Basically in RVMODEL_BOOT you could
> > have
> > > a label : rvmodel_boot_start and use this as the entry point in the
> > linker
> > > script. Have your boot-up code and then simply fall into
> > RVTEST_CODE_BEGIN.
> > >
> > > Essentially replace rvtest_start in this line
> > > <
> >
> https://github.com/incoresemi/riscv-compliance/blob/34d979eaf859d8608adaa4781235f327bc51a5f8/riscv-target/spike/link.ld#L2
> > >
> > > with rvmodel_boot_start.
> > >
> > > So then you don't need RVMODEL_TARGET_INIT.
> > >
> > > —
> > > You are receiving this because you authored the thread.
> > > Reply to this email directly, view it on GitHub
> > > <
> >
> https://github.com/incoresemi/riscv-compliance/pull/1#discussion_r520691808
> > >,
> > > or unsubscribe
> > > <
> >
> https://github.com/notifications/unsubscribe-auth/AD72ZM533VADHHOPOLYI6LLSPFR5LANCNFSM4TM53OMA
> > >
> > > .
> > >
> >
> >
> > --
> > /****************
> > Marc Karasek
> > ASM T648
> > Herder of Cats
> > ****************/
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/incoresemi/riscv-compliance/pull/1#issuecomment-724995909
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AHPXVJUXYLSL6ASDG6ZJF53SPG2QVANCNFSM4TM53OMA
> >
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/incoresemi/riscv-compliance/pull/1#issuecomment-725005188>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD72ZM6R52WSTX5CT25P4B3SPG44TANCNFSM4TM53OMA>
> .
>


-- 
/****************
Marc Karasek
ASM T648
Herder of Cats
****************/

neelgala commented:
@MarcKarasek 

Considering that none of the targets in the repo have a boot-code or a requirement for it, I would say its not a common case. So if someone was requiring a boot-code, a little effort to change the linker script wouldn't hurt - doesn't affect the masses. Again, this is a tiny subset for me to generalize on and can be easily argued against so i'll leave it at that.

I agree the spec needs more clarity in regards to the linker script and more information on the use of RVMODEL_BOOT. Will see if I can update the spec on this with @allenjbaum's help.

Also other than the "ease of use" argument I hope there is no technical reasons restricting to have RVMODEL_BOOT before RVTEST_CODE_BEGIN and changing the label in the linker script.

I like that we narrowed down the issue to having a fixed entry point. If we do decide on it my suggestion would be to subsume the RVMODEL_BOOT inside the RVTEST_CODE_BEGIN with an explicit rvtest_entry_point label before rvtest_start. The revised version would look like this:

```
.macro RVTEST_CODE_BEGIN
  .align UNROLLSZ
  .section .text.init;
  rvtest_entry_point:
    RVMODEL_BOOT
  .option norelax;
  .globl rvtest_start;                                                  
  rvtest_start:
```
This would keep the entry point constant for all linker scripts and also avoid BOOT-CODE from being part of the actual test itself (which is used for coverage and data propagation analysis).

Thoughts ?

note, the linker is still target specific, and a user could always change the entry point to be anywhere.


PS: I think we should chage rvtest_start to rvtest_init since it includes test-initialization routine. The actual test cases only start after rvtest_code_begin label.

MarcKarasek commented:
I can add to the spec my sample ld script and explanation i sent out to the
mailing list.

So the example ld script would have to have rvtest_entry_point as its start
execution address..
And RVMODEL_BOOT would be removed from all the tests.



On Tue, Nov 10, 2020 at 11:49 PM Neel Gala <notifications@github.com> wrote:

> @MarcKarasek <https://github.com/MarcKarasek>
>
> Considering that none of the targets in the repo have a boot-code or a
> requirement for it, I would say its not a common case. So if someone was
> requiring a boot-code, a little effort to change the linker script wouldn't
> hurt - doesn't affect the masses. Again, this is a tiny subset for me to
> generalize on and can be easily argued against so i'll leave it at that.
>
> I agree the spec needs more clarity in regards to the linker script and
> more information on the use of RVMODEL_BOOT. Will see if I can update the
> spec on this with @allenjbaum <https://github.com/allenjbaum>'s help.
>
> Also other than the "ease of use" argument I hope there is no technical
> reasons restricting to have RVMODEL_BOOT before RVTEST_CODE_BEGIN and
> changing the label in the linker script.
>
> I like that we narrowed down the issue to having a fixed entry point. If
> we do decide on it my suggestion would be to subsume the RVMODEL_BOOT
> inside the RVTEST_CODE_BEGIN with an explicit rvtest_entry_point label
> before rvtest_start. The revised version would look like this:
>
> .macro RVTEST_CODE_BEGIN
>   .align UNROLLSZ
>   .section .text.init;
>   rvtest_entry_point:
>     RVMODEL_BOOT
>   .option norelax;
>   .globl rvtest_start;
>   rvtest_start:
>
> This would keep the entry point constant for all linker scripts and also
> avoid BOOT-CODE from being part of the actual test itself (which is used
> for coverage and data propagation analysis).
>
> Thoughts ?
>
> note, the linker is still target specific, and a user could always change
> the entry point to be anywhere.
>
> PS: I think we should chage rvtest_start to rvtest_init since it includes
> test-initialization routine. The actual test cases only start after
> rvtest_code_begin label.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/incoresemi/riscv-compliance/pull/1#issuecomment-725196279>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD72ZM2JHR3TXBAOCEQ52YTSPIJVNANCNFSM4TM53OMA>
> .
>


-- 
/****************
Marc Karasek
ASM T648
Herder of Cats
****************/

neelgala commented:
@MarcKarasek @allenjbaum all looks good to me. If you agree, can I go ahead and merge this ?

MarcKarasek commented:
Yes

Sent from Nexus 7
Marc Karasek
92-57 Bear
Troop 648 Scoutmaster

On Mon, Nov 16, 2020, 11:34 PM Neel Gala <notifications@github.com> wrote:

> @MarcKarasek <https://github.com/MarcKarasek> @allenjbaum
> <https://github.com/allenjbaum> all looks good to me. If you agree, can I
> go ahead and merge this ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/incoresemi/riscv-compliance/pull/1#issuecomment-728679888>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD72ZM3BP3QY5SHHEBTJ663SQH4MZANCNFSM4TM53OMA>
> .
>
